dataset.py

```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
 
   return PartitionedDataset(train=train_data, test=test_data, validation=validation_data)
 
+def _nearest_neighbors(
+  center_coordinate: list[float],
+  sample_data: pd.DataFrame,
+  n_neighbors: int) -> pd.DataFrame:
+  '''
+  From sample_data, select the n_neighbors closest to the specified
+  center_coordinate.
+  '''
+  assert(len(center_coordinate) == 2)
+  # Sometimes rounding of split fractions with dataframe shapes
+  # may end up with bigger or smaller splits.
+  n_neighbors = int(
+    min(sample_data.shape[0], n_neighbors))
+  if n_neighbors < 3:
+    return pd.DataFrame({
+      _LONGITUDE_COLUMN_NAME: [],
+      _LATITUDE_COLUMN_NAME: [],
+    })
+  near = NearestNeighbors(n_neighbors=n_neighbors).fit(
+    sample_data[[_LONGITUDE_COLUMN_NAME, _LATITUDE_COLUMN_NAME]]
+  )
+  indices = near.kneighbors(
+    [center_coordinate], return_distance=False)[0]
+  return sample_data.iloc[indices, :]
+
+def _partition_by_nearest_neighbors(
+  sample_data: pd.DataFrame,
+  strategy: FurthestPointsPartitionStrategy,
+  train_coord: tuple,
+  validation_coord: tuple,
+  test_coord: tuple
+):
+  '''
+  Returns a PartitionedDataset from splits of a sample_data pd.DataFrame by the train,
+  validation and test split fractions, choosing the points based on the nearest points
+  for each of the train_coord, validation_coord and test_coord arguments.
+  It first gets the nearest neighbors for train_coord, then validation_coord
+  and finally test_coord.
+  '''
+  train_split = _nearest_neighbors(
+    center_coordinate=train_coord,
+    sample_data=sample_data,
+    n_neighbors=sample_data.shape[0]*strategy.train_fraction
+  )
+  filtered_sample_data = sample_data[
+          ~sample_data.index.isin(
+            train_split.index.values
+          )]
+  
+  validation_split = _nearest_neighbors(
+    center_coordinate=validation_coord,
+    sample_data=filtered_sample_data,
+    n_neighbors=sample_data.shape[0]*strategy.validation_fraction 
+  )
+  filtered_sample_data = filtered_sample_data[
+    ~filtered_sample_data.index.isin(
```

Comment:
This is a bitwise operator. I think we probably want the logical NOT operator `not`

Reply:
Within pandas dataframe operations, ~ is the required operator. NOT is not compatible.

See:
https://screenshot.googleplex.com/AThFnL3WAKNK2nU
https://sparkbyexamples.com/pandas/not-in-filter-in-pandas/